iarc_roombafilter/scripts/f_ros.py

```diff
@@ -0,0 +1,170 @@
+    def __init__(self, dt):
+
+        # unroll parameters ...
+        self._max_targets = rospy.get_param('~max_targets', default=30)
+        self._sigma = np.copy(rospy.get_param('~sigma', default=list(cfg.SIGMAS)))
+        Particle.set_sigmas(np.copy(self._sigma))
+        self._p_keep = rospy.get_param('~p_keep', default=cfg.P_KEEP)
+        self._p_match = rospy.get_param('~p_match', default=cfg.P_MATCH)
+        self._p_clear = rospy.get_param('~p_clear', default=cfg.P_CLEAR)
+        self._obs_topic = rospy.get_param('~obs_topic', default='visible_roombas')
+        self._out_topic = rospy.get_param('~out_topic', default='filtered_roombas')
+        self._rate = rospy.get_param('~rate', default=50) #50hz
+
+        # create ROS interfaces
+        self._tf = tf.TransformListener(True, cache_time=rospy.Duration(5))
+        self._sub = rospy.Subscriber(self._obs_topic, RoombaSighting, self.obs_cb, queue_size=1)
+        self._pub = rospy.Publisher(self._out_topic, RoombaList, queue_size=10)
+        self._mgr = UKFManager(dt, self._sigma)
+        self._t = rospy.Time.now()
+
+        self._obs = None
+
+    def obs_cb(self, msg):
+        """
+        Save observation for processing ...
+        """
+        self._obs = msg
+
+    def obs_proc(self):
+        """
+        Perform filtering according to observation.
+        Decoupled from obs_cb to only process latest observation.
+        """
+        msg = self._obs
+        self._obs = None
+        if msg == None:
+            return
+
+        try:
+            # TODO : kind-of ugly hack to pull stamp into most recent available tf
+            # better way?
+            msg.fov_center.header.stamp = rospy.Time(0)
+            pos = self._tf.transformPoint('map', msg.fov_center)
+        except tf.Exception as e:
+            return
+
+        obs_ar = CircularObservation(
+                pos.point.x,
+                pos.point.y,
+                msg.fov_radius
+                )
+
+        # format observation data ...
+        obs = []
+        for r in msg.data:
+
+            #print r.visible_location.header.frame_id
+            #print r.visible_location.pose.pose.position.x
+
+            p = PoseStamped(r.visible_location.header, r.visible_location.pose.pose)
+            p = self._tf.transformPose('map', p).pose
+            p, q = p.position, p.orientation
+            h = tf.transformations.euler_from_quaternion([q.x,q.y,q.z,q.w])[2]
+
+            pose = np.asarray([p.x, p.y, h])
+            # note : no velocity information is available.
+
+            # roomba type handling
+            r_type = cfg.T_NULL
+            if r.type != Roomba.UNKNOWN:
+                if r.type == Roomba.OBSTACLE:
+                    r_type = cfg.T_OBST
+                else:
+                    r_type = cfg.T_RED if r.type == Roomba.RED else cfg.T_GREEN
+
+            o = ObservationParticle(
+                    pose=pose,
+                    t=r_type,
+                    )
```

Comment:
Should this include somehow the covariance from the observation?

Reply:
added a todo.